targets/TARGET_STM/qspi_api.c

```diff
@@ -25,7 +25,12 @@
 #include "pinmap.h"
 #include "PeripheralPins.h"
 
+#if MBED_CONF_MBED_TRACE_ENABLE
 #include "mbed-trace/mbed_trace.h"
+#else
+#define tr_info(...)
+#define tr_error(...)
+#endif // MBED_CONF_MBED_TRACE_ENABLE
```

Comment:
In the long run, I personally think we should avoid using `mbed-trace` altogether in `platform`, `drivers` and `targets` which are bare-minimal parts of mbed-os available to the bare metal profile by default.

Reply:
mbed trace is fundamental enough that surely it should be in platform, or at least be included as a component by default, in which case it would provide its own dummy definitions.

It doesn't make sense to start scattering local dummy definitions everywhere,

Reply:
Tracing should be no-op if disabled but available for everyone (part of platform).

Reply:
@0xc0170 `mbed_trace.h` already contains dummy definitions (no-op), this is correct. The `mbed-trace` library isn't available to bare metal (unless the user adds it to `mbed_app.json`), and `#include "mbed-trace/mbed_trace.h"` fails.

One option is to add `mbed-trace` to https://github.com/ARMmbed/mbed-os/blob/0db72d0cf26539016efbe38f80d6f2cb7a3d4414/platform/bare_metal/mbed_lib.json#L1-L4
@kjbracey-arm said trace is fundamental enough, so shall we have it by default for bare metal? @evedon

Reply:
By default tracing is disabled and so there is no memory cost for it. I agree that it is fundamental enough to be included by default in bare metal.

Reply:
Once updated, let us know to start CI

Reply:
I think this PR has to be closed,
and a new one will affect platform side ?

Reply:
I tried to do the change on the platform side (e.g. making mbed-trace default), but stomped on a new issue. `mbed-trace` has ipv6 enabled by default, which further depends on `mbed-client-libservice` (a network library)...
https://github.com/ARMmbed/mbed-os/blob/dcf6def9478b7ab85762acfdede935131e2e9e97/platform/mbed-trace/source/mbed_trace.c#L24-L32
The default value is undefined (null):
https://github.com/ARMmbed/mbed-os/blob/dcf6def9478b7ab85762acfdede935131e2e9e97/platform/mbed-trace/mbed_lib.json#L14-L17
And the source file treats undefined as 1 (enabled) as you can see.

But, if I try to fix it and make it disabled by default, it'll impact existing users who assume ipv6 tracing to be on...

Reply:
You could disable ipv6 tracing in bare metal builds. Similar to what we do in the test config https://github.com/ARMmbed/mbed-os/blob/master/TESTS/configs/baremetal.json/#L41

Reply:
Good idea, as network stack has RTOS dependencies and can't be enabled by users in bare metal builds

Reply:
@evedon Unfortunately, it seems that we can't override `mbed-trace.fea-ipv6` in `platform/bare_metal/mbed_lib.json`. As far as I tested, a library can only override its own configs + `target.*` configs, but not configs of another library. @0xc0170 Could you confirm this?

And we can't resort to checking macros, because
* `bare-metal` doesn't define any macros
* `mbed-client-libservice` (where ipv6 comes from) doesn't define any macros
* `MBED_CONF_RTOS_API_PRESENT` is what we normally use. But checking RTOS for ipv6 in the code may cause confusion.

Reply:
>As far as I tested, a library can only override its own configs + target.* configs, but not configs of another library. @0xc0170 Could you confirm this?

I don't know but it would make sense if only a user could change it.

> mbed-client-libservice (where ipv6 comes from) doesn't define any macros

Can it define macros -  would that allow us to fix this?

Reply:
You can add a dummy "present" option to the mbed_lib.json for nanostack-libservice, then check its macro (`MBED_CONF_NANOSTACK_LIBSERVICE_PRESENT`). That's how we detect RTOS or RTOS API presence.

Then `fea-ipv6` = `true`could only take effect if that is turned on too.

Or, even more fancy, change the default for `fea-ipv6` to `null`, which is interpreted as "on if libservice present". If the user explicitly sets it to `true`, then you can give an error telling them to add libservice if necessary.

Reply:
I can add that config to `mbed-client-libservice` (part of https://github.com/ARMmbed/nanostack-libservice), but this needs to be communicated with the client team who fetch libraries from standalone repos sometimes. Maybe the change needs to be "upstream".

> Or, even more fancy, change the default for `fea-ipv6` to `null`, which is interpreted as "on if libservice present".

It is already null 😉

Reply:
Okay, yes, there are a few wrinkles there. The header defaults to IPv6 on if the `MBED_CONF_MBED_TRACE_FEA_IPV6` is unset, which it is for `null` in Mbed OS, or if not using Mbed OS config.

You could distinguish that further by `TARGET_LIKE_MBED` - if that's defined, then check the `MBED_CONF_NANOSTACK_LIBSERVICE_PRESENT`, else assume on. 

Still, could have a version mismatch where they get a new mbed trace from upstream but use it with old Mbed OS. Not sure if that's a possibility.

In non-Mbed OS builds, libservice is always there - it's Mbed OS that's being the odd thing out by potentially having mbed trace without libservice. Would it be the end of the world to add the dependency and add 10(?) more small source files to the bare build? Still shouldn't add code.

> It is already null

Hmph. There's a bad habit of making boolean options be `null`/`true`, when they should be `false`/`true`. `null` should only be used for "none" (which would normally be in place of a number or string), or "default" - which could apply to booleans, but is only worth doing if it's some sort of "dynamic" setting, not just a constant true or false.

Reply:
I'm going to make a PR with the minimum amount of check, and ping the client team.

> Hmph. There's a bad habit of making boolean options be `null`/`true`, when they should be `false`/`true`. `null` should only be used for "none" (which would normally be in place of a number or string), or "default" - which could apply to booleans, but is only worth doing if it's some sort of "dynamic" setting, not just a constant true or false.

In fact, the code respects `true`/`false` set by users, and treats `null` as `true` otherwise. So now it's super easy for us to add the check for nanostack-libservice...

Reply:
#13649 created